process/throttle/block.go

```diff
@@ -0,0 +1,144 @@
+type blockSizeThrottle struct {
+	statistics   []*blockInfo
+	maxItems     uint32
+	mutThrottler sync.RWMutex
+}
+
+// NewBlockSizeThrottle creates a new blockSizeThrottle object
+func NewBlockSizeThrottle() (*blockSizeThrottle, error) {
+	bst := blockSizeThrottle{}
+	bst.statistics = make([]*blockInfo, 0)
+	bst.maxItems = process.MaxItemsInBlock
+	return &bst, nil
+}
+
+// MaxItemsToAdd gets the maximum number of items which could be added in one block, taking into consideration the
+// previous results
+func (bst *blockSizeThrottle) MaxItemsToAdd() uint32 {
+	bst.mutThrottler.RLock()
+	maxItems := bst.maxItems
+	bst.mutThrottler.RUnlock()
+
+	return maxItems
+}
+
+// Add adds the new state for last block which has been sent and updates the max items which could be added
+// in one block
+func (bst *blockSizeThrottle) Add(round uint64, items uint32) {
+	bst.mutThrottler.Lock()
+	bst.statistics = append(
+		bst.statistics,
+		&blockInfo{round: round, items: items, maxItems: bst.maxItems},
+	)
+	bst.mutThrottler.Unlock()
+}
+
+// Succeed sets the state of the last block which has been sent at the given round
+func (bst *blockSizeThrottle) Succeed(round uint64) {
+	bst.mutThrottler.Lock()
+	for i := len(bst.statistics) - 1; i > 0; i-- {
+		if bst.statistics[i].round == round {
+			bst.statistics[i].succeed = true
+			break
+		}
+	}
+	bst.mutThrottler.Unlock()
+}
+
+// ComputeMaxItems computes the max items which could be added in one block, taking into consideration the previous
+// results
+func (bst *blockSizeThrottle) ComputeMaxItems() {
+	//TODO: This algorithm is now something basic and could be improved
+	bst.mutThrottler.Lock()
+	defer func() {
+		log.Info(fmt.Sprintf("max number of items which could be added in one block is %d\n", bst.maxItems))
+		bst.mutThrottler.Unlock()
+	}()
+
+	if len(bst.statistics) == 0 {
+		return
+	}
+
+	lastActionSucceed := bst.statistics[len(bst.statistics)-1].succeed
+	lastActionMaxItems := bst.statistics[len(bst.statistics)-1].maxItems
+
+	if lastActionSucceed {
+		bst.maxItems = bst.getMaxItemsWhenSucceed(lastActionMaxItems)
+	} else {
+		bst.maxItems = bst.getMaxItemsWhenNotSucceed(lastActionMaxItems)
+	}
+}
+
+func (bst *blockSizeThrottle) getMaxItemsWhenSucceed(lastActionMaxItems uint32) uint32 {
+	if lastActionMaxItems == process.MaxItemsInBlock {
+		return lastActionMaxItems
+	}
+
+	noOfMaxItemsUsedWithoutSucceed := bst.getCloserAboveMaxItemsUsedWithoutSucceed(lastActionMaxItems)
+	if lastActionMaxItems*100/noOfMaxItemsUsedWithoutSucceed > jumpAbovePercent {
+		return noOfMaxItemsUsedWithoutSucceed
+	}
+
+	increasedNoOfItems := core.Max(1, uint32(float32(noOfMaxItemsUsedWithoutSucceed-lastActionMaxItems)*jumpAboveFactor))
+	return lastActionMaxItems + increasedNoOfItems
+}
+
+func (bst *blockSizeThrottle) getCloserAboveMaxItemsUsedWithoutSucceed(currentMaxItems uint32) uint32 {
+	for i := len(bst.statistics) - 1; i > 0; i-- {
```

Comment:
Same here, shouldn't it be i>=0?

Reply:
Yep. I already fixed it when I have added units tests.